url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/CPU.java

```diff
@@ -0,0 +1,491 @@
+    }
+    long next_print = cycles + CYCLES_PER_DEBUG;
+    // How much should this be???
+    monitor.info("Starting CPU at: " + Integer.toHexString(pc));
+    try {
+      while (running) {
+
+        // Debugging?
+        if (monitor.isEnabled()) { // || interruptInExec > 0) {
+          if (baLowUntil <= cycles) {
+            fixRindex(pc); // sets the rindex!
+            monitor.disAssemble(memory, rindex, acc, x, y,
+                (byte) getStatusByte(), interruptInExec,
+                lastInterrupt);
+          }
+        }
+
+        // Run one instruction!
+        emulateOp();
+
+        // Also allow the 1541 to run an instruction!
+        if (EMULATE_1541) {
+          c1541.tick(cycles);
+        }
+
+        nr_ins++;
+        if (next_print < cycles) {
+          long sec = System.currentTimeMillis() - lastMillis;
+          int level = monitor.getLevel();
+
+          if (DEBUG && level > 1) {
+            monitor.info("--------------------------");
+            monitor.info("Nr ins:" + nr_ins + " sec:" + (sec)
+                + " -> " + ((nr_ins * 1000) / sec) + " ins/s"
+                + "  " + " clk: " + cycles + " clk/s: "
+                + ((CYCLES_PER_DEBUG * 1000) / sec) + "\n"
+                + ((nr_irq * 1000) / sec));
+            if (level > 2)
+              monitor.disAssemble(memory, rindex, acc, x, y,
+                  (byte) getStatusByte(), interruptInExec,
+                  lastInterrupt);
+            monitor.info("--------------------------");
+          }
+          nr_irq = 0;
+          nr_ins = 0;
+          lastMillis = System.currentTimeMillis();
+          next_print = cycles + CYCLES_PER_DEBUG;
+        }
+      }
+    } catch (Exception e) {
+      monitor.error("Exception in loop " + pc + " : " + e);
+      e.printStackTrace();
+      monitor.disAssemble(memory, rindex, acc, x, y,
+          (byte) getStatusByte(), interruptInExec, lastInterrupt);
+    }
+  }
+
+  // -------------------------------------------------------------------
+  // Cheat loop!
+  // Protection
+  // + rule triggered auto get/store
+  // Rule: xpr & xpr & xpr ...
+  // rule: int[] adr, cmptype, cmpval ...
+  // autostore: int[] adr, len => result in hex! from adr and on!
+  // -------------------------------------------------------------------
+
+  public void setAutoStore(int index, AutoStore au) {
+    autoStore[index] = au;
+  }
+
+  public AutoStore getAutoStore(int index) {
+    return autoStore[index];
+  }
+
+  public void setCheatEnabled(int maxAutostores) {
+    cheatMon = new int[0x10000];
+    autoStore = new AutoStore[maxAutostores];
+  }
+
+  public void protect(int address, int value) {
+    cheatMon[address] = (cheatMon[address] & 0xff) | (value << 8) | CH_PROTECT;
+  }
+
+  public void monitorRead(int address) {
+    cheatMon[address] |= CH_MONITOR_READ;
+  }
+
+  public void monitorWrite(int address) {
+    cheatMon[address] |= CH_MONITOR_WRITE;
+  }
+
+
+  public void cheatLoop() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.